quote-book/synchronized/src/lib.rs

```diff
@@ -1,111 +1,392 @@
+    }
+
+    fn num_scis(&self) -> Result<u64, QuoteBookError> {
+        self.synchronized_quote_book_impl.num_scis()
+    }
+}
+
+/// A callback for broadcasting a quote removal. It receives 1 argument:
+/// - A vector of quotes that have been removed
+pub type RemoveQuoteCallback = Arc<Mutex<dyn FnMut(Vec<Quote>) + Sync + Send>>;
+
+struct DbFetcherThread<DB: Ledger, Q: QuoteBook> {
+    db: DB,
+    quotebook: Q,
+    /// Message bus sender.
+    remove_quote_callback: RemoveQuoteCallback,
+    next_block_index: u64,
+    highest_processed_block_index: Arc<AtomicU64>,
+    stop_requested: Arc<AtomicBool>,
+    logger: Logger,
+}
+
+/// Background worker thread implementation that takes care of periodically
+/// polling data out of the database. Add join handle
+impl<DB: Ledger, Q: QuoteBook> DbFetcherThread<DB, Q> {
+    const POLLING_FREQUENCY: Duration = Duration::from_millis(10);
+    const ERROR_RETRY_FREQUENCY: Duration = Duration::from_millis(1000);
+
+    pub fn start(
+        db: DB,
+        quotebook: Q,
+        remove_quote_callback: RemoveQuoteCallback,
+        highest_processed_block_index: Arc<AtomicU64>,
+        stop_requested: Arc<AtomicBool>,
+        logger: Logger,
+    ) {
+        let next_block_index = highest_processed_block_index.load(Ordering::SeqCst);
+        let thread = Self {
+            db,
+            quotebook,
+            remove_quote_callback,
+            next_block_index,
+            highest_processed_block_index,
+            stop_requested,
+            logger,
+        };
+        thread.run();
+    }
+
+    fn run(mut self) {
+        log::info!(self.logger, "Db fetcher thread started.");
+        loop {
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Db fetcher thread stop requested.");
+                break;
+            }
+            let starting_block_index = self.next_block_index;
+            while self.load_block_data() && !self.stop_requested.load(Ordering::SeqCst) {}
+            if self.next_block_index != starting_block_index {
+                let last_processed_block_index = self.next_block_index - 1;
+                let mut quotes = self
+                    .quotebook
+                    .remove_quotes_by_tombstone_block(last_processed_block_index);
+                while quotes.is_err() {
+                    log::error!(
+                        self.logger,
+                        "Unexpected error when removing quotes by tombstone_block {}. Retrying. Error: {:?}",
+                        last_processed_block_index,
+                        quotes.err()
+                    );
+                    std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    quotes = self
+                        .quotebook
+                        .remove_quotes_by_tombstone_block(last_processed_block_index);
+                }
+                (self.remove_quote_callback.lock().expect("lock poisoned"))(quotes.unwrap());
```

Comment:
what are we unwrapping here? this looks very fishy

Reply:
i see, it's a result, but you know that `is_err` is false. maybe we can change this to `.expect` and put a message to that effect